### DISCUSSION TITLE
bosh-release 271  Inappropriate ioctl for device (Errno::ENOTTY)

### DISCUSSION THREAD

**Discussion by saketlj** [cloudfoundry/bosh#2351] [General]
Hi Team,

We are using bosh-release 271 and logging to bosh console with /var/vcap/jobs/director/bin/console and running some ruby inbuilt bosh code to clean up DB tasks.

count=(Bosh::Director::Models::Task.where(type: 'vms').count-2000)/10

I am getting below error regards to ioctl winsize error.

/var/vcap/data/packages/ruby-3.0.2-r0.67.0/23d7612192d5eb20127684cc602ad13a0778fdc1/lib/ruby/3.0.0/irb/input-method.rb:42:in `winsize': Inappropriate ioctl for device (Errno::ENOTTY)

Inappropriate ioctl for device normally means that some program attempted to do a terminal control operation but its standard I/O streams weren't connected to a terminal.

Did any 1 came across this error ?

**Comment by beyhan** [comment]
Hi @saketlj,

I can't reproduce the error above with the bosh version `271.16.0`. Which version do you use exactly?

**Reply by saketlj** [reply]
@beyhan I am using 271.8.0 bosh release.

Strange thing is when i run this manually ( login to bosh director and the calling /var/vcap/jobs/director/bin/console) I am able to run this successfully, only first attempt was unsuccessful with same error.

I am doing this through automated flow with concourse task, copying bash script to director which runs below code.

https://community.pivotal.io/s/article/How-to-clean-up-stale-BOSH-tasks-history-from-console?language=en_US&t=1641995718671

#/bin/bash
/var/vcap/jobs/director/bin/console <<EOF
count=(Bosh::Director::Models::Task.where(type: 'vms').count-2000)/10
if count > 0
    tr=Bosh::Director::Api::TaskRemover.new(2000)
    count.times {tr.remove('vms')}
end
EOF

**Reply by saketlj** [reply]
This seems ruby issue.
https://stackoverflow.com/questions/64934029/error-in-rails-console-when-interacting-with-database-tables
https://youtrack.jetbrains.com/issue/RUBY-26249

**Reply by saketlj** [reply]
@beyhan We are still having this issue and bosh version is 271.20.0

When I run script manually it runs successfully but when I try to redirect o/p of this script to temp file, it prints below ioctl error message after variable assignment operation only.

for e.g :
count = Bosh::Director::Models::Task.where(type: 'vms').count-1000
/var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/input-method.rb:42:in `winsize': Inappropriate ioctl for device (Errno::ENOTTY)
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/input-method.rb:42:in `winsize'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:800:in `output_value'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:572:in `block (2 levels) in eval_input'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:751:in `signal_status'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:548:in `block in eval_input'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/ruby-lex.rb:251:in `block (2 levels) in each_top_level_statement'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `loop'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `block in each_top_level_statement'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `catch'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `each_top_level_statement'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:547:in `eval_input'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:481:in `block in run'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:480:in `catch'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:480:in `run'
	from /var/vcap/data/packages/ruby-3.0.2-r0.67.0/d1fad89d06d431d270f10b89bb1bac4e44e64cea/lib/ruby/3.0.0/irb.rb:409:in `start'
	from /var/vcap/data/packages/director/48ac26155a037b2d2ebf2d04d7fe7c269dbdf9d9/gem_home/ruby/3.0.0/gems/bosh-director-0.0.0/bin/bosh-director-console:44:in `start'
	from /var/vcap/data/packages/director/48ac26155a037b2d2ebf2d04d7fe7c269dbdf9d9/gem_home/ruby/3.0.0/gems/bosh-director-0.0.0/bin/bosh-director-console:75:in `<top (required)>'
	from /var/vcap/packages/director/bin/bosh-director-console:29:in `load'
	from /var/vcap/packages/director/bin/bosh-director-console:29:in `<main>'
Maybe IRB bug!

I haven't found any solution in ruby community for this.

**Comment by beyhan** [comment]
@saketlj This feature is there only for debugging purpose as described in the [docs](https://bosh.io/docs/tips/#director-db). What you do isn't a supported use case for this. We are happy to review a pr about his if you want to contribute.
